cprogs/summer.1
.TH SUMMER "1" "December 2006" "Debian" "Chiark-utils-bin"
.SH NAME
summer \- print checksum and system metainformation for files
.SH SYNOPSIS
.B summer -ACDbfqtx
.RI [\| startpoint ...]
.br
.SH DESCRIPTION
.B summer
prints the MD5 checksum of the contents, and the system
metainformation (ownership, permissions, timestamps, etc.), for a
file, or recursively for a whole directory tree.

Each command line argument should be a file or directory to be processed;
if it is a directory then it will be processed and then its contents will
also be processed, recursively. If no
.IR startpoint s
are specified on the command line then
.B summer
will read a list of newline-separated startpoints from standard input.

Since
.B summer
correctly handles devices, FIFOs and other non-regular files it is useful
for generating and comparing summaries of arbitrary directory trees where
md5sum alone would not be.
.SH OUTPUT FORMAT
.B summer
prints one line of information for each filesystem object it processes.
Each line has the following columns:
.TS
tab (@);
l l.
@MD5 checksum (in hex) or file type information
@Size of file in bytes
@File access rights (in octal)
@User ID of owner (in decimal)
@Group ID of owner (in decimal)
@atime (time of last access, decimal time_t)
@mtime (time of last modification)
@ctime (time of last status change)
@Filename
.TE

For regular files, the first column is the md5sum. For directories, pipes,
symlinks and sockets it is the literal string \fBdir\fR, \fBmountpoint\fR, \fBpipe\fR, \fBsymlink\fR or \fBsocket\fR
as appropriate. For devices it begins with \fBc\fR for character or \fBb\fR for block
devices, followed by the device number as a single 32 bit hex number and as
four separate 8 bit decimal numbers (most significant first).

Note that any bytes in the filename other than printing 7-bit ASCII
are escaped using
.B \\\x\c
.I NN
syntax, where
.I NN
are two hex digits; backslashes are also escaped in this way.
This makes the output unambiguous.  Filenames will be relative
if the relevant
.I startpoint
was relative, and absolute if it was absolute.

For symlinks the filename column is followed by `\fB -> \fR' (note the
spaces) and the target of the link, again escaped, as above.
.SH OPTIONS
.TP
.B \-A
Do not print the atime (time of last access). The atime column will be omitted.
.TP
.B \-C
Do not print the ctime (time of last status change). The ctime column will be omitted.
.TP
.B \-M
Do not print the mtime (time of last modification). The mtime column will be omitted.
.TP
.B \-D
Do not print directory sizes. The size column for directories will read \fBdir\fR.
.TP
.B \-b
Do not print mtime (time of last modification) for symbolic links. The mtime field
for symbolic links will read \fBlink\fR.
.TP
.B \-B
Do not print any times for special files, symlinks,
sockets, or fifos.  The atime, mtime and ctime fields
for these objects will read
.BR char ", " block ", " link  ", " sock " or " pipe
instead.
.TP
.B \-f
Include information about errors encountered (for example, unreadable files)
in the output, and continue processing. The default is to print error information
to standard error and stop immediately an error is encountered.
.TP
.B \-x
Do not cross mountpoints while recursing into subdirectories.  
Startpoints which are mountpoints \fIare\fR descended into.
.TP
.B \-q
Suppress the progress information which
.B summer
normally prints to standard error.
.TP
.B \-t
Set the field separator between the information and the filename to a
tab character (default is space).
.TP
.B \-f
Normally any errors (problems accessing files including nonexistent
startpoings, and the like) are fatal; an error message is reported to
stderr.

With
.B -f
errors are nonfatal and the problems are reported inline.  The
filesystem object with the problem is reported in the normal way
except that instead of the checksum, the string
\fB\\[\fR\fIproblem\fR[\fB:\fR\ \fIdetails\fR]\fB]\fR
appears.  Fields whose value could not be determined are printed
as \fB?\fR.
.TP
.B \-h
Print a brief usage message to stderr (and do nothing else, exiting nonzero).
.SH PARSING THE OUTPUT
If the first character in the line is \fB\\[\fR, then the first
(checksum or type) field is everything until the first subsequent
\fB]\fR; this may be of variable length and will be followed by one or
more spaces.  Otherwise the first field has a fixed width: 64
characters, the size of an MD5 checksum represented in hex, and is
followed by a single space.

The metadata fields are space-separated but are also space-padded to a
minimum width: 10 characters for sizes and times and ids; 4 characters
for the mode.

The filename field, and optional link target information, are of
variable length, but they are escaped so that they do not contain
spaces.
.SH AUTHOR
.B summer
is
.br
Copyright (C) 2003-2007 Ian Jackson <ian@chiark.greenend.org.uk>

This manpage was written by Peter Maydell
and subsequently improved by Ian Jackson.  It is
.br
Copyright (C) 2006 Peter Maydell <pmaydell@chiark.greenend.org.uk>
.br
Copyright (C) 2007 Ian Jackson <ian@chiark.greenend.org.uk>

This is free software, distributed under the GNU General Public
Licence, version 3 or (at your option) any later version; see
/usr/share/doc/chiark-utils-bin/copyright or
/usr/share/common-licenses/GPL-3
for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


debian/changelog
chiark-utils (8.0.0-0deepin1) unstable; urgency=medium

  * feat: add sw64 support 

 -- hudeng <hudeng@deepin.org>  Tue, 21 Oct 2025 17:32:41 +0800

chiark-utils (8.0.0) unstable; urgency=medium

  * Rewrite the packaging to use `debhelper` and `dh` in as standard a way
    as possible.  As a consequence, support root-less builds.
    No functional changes to the installed packages.
    Closes: #1089303.  [Niels Thykier]

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Sun, 12 Jan 2025 20:19:58 +0000

chiark-utils (7.0.3) unstable; urgency=medium

  * Team upload.
  * git-cache-proxy: Handle extra-parameters in git protocol (Closes: #1040476).
    Thanks to Anthony PERARD for the patch.
  * git-branchmove: Always pass --no-verify to 'git push' (Closes: #1080082).

 -- Sean Whitton <spwhitton@spwhitton.name>  Thu, 26 Sep 2024 17:07:41 +0100

chiark-utils (7.0.2) unstable; urgency=medium

  * debian/rules: Pass -D_TIME_BITS=64 too.  Closes: #1078190.
  * Change file size type to unsigned long long.  Closes: #1075821.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Thu, 15 Aug 2024 22:28:46 +0100

chiark-utils (7.0.1) unstable; urgency=medium

  * backup/snaprsync: add -AX to the rsync rune
  * Enable LFS and 64-bit time_t support.  Closes: #1067157.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Thu, 04 Jul 2024 12:53:58 +0100

chiark-utils (7.0.0) unstable; urgency=medium

  New facility:
  * prefork-interp, Proc::Prefork::Interp: new facility

  Improvements:
  * expire-iso8601: Replace with new implementation.  Closes:#862897.

  Bugfixes:
  * acctdump: Fix manual page section.
  * git-cache-proxy: fix for new git.  Closes: #1006695.  [Anthony Perard]
  * git-branchmove: install manpage..  Closes: #946364.  [Sean Whitton]
  * hexterm: Update #! and Suggest tcl8.6.  Closes: #905474.

  Build and packaging improvements:
  * Fix cross build.  Closes: #951242.  [Helmut Grohne]
  * copyright: Be clear that SPDX indications are licence grants
  * Add lintian override for python dependency
  * Add lintian override for chiark-really
  * fix priority "extra" to "optional"
  * Fix build-dependency on debhelper.
  * Set source format to 3.0 (native)
  * Add Vcs-* fields.
  * Update Standards-Version
  * timespeccmp.h: Move use to just cgi-fcgi-interp

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 22 Aug 2022 00:06:57 +0100

chiark-utils (6.1.2+nmu1) unstable; urgency=medium

  * Non-maintainer upload.
    - Maintainer has approved this via personal communication.
  * Bump debhelper compat to 13 (Closes: #965453).

 -- Sean Whitton <spwhitton@spwhitton.name>  Tue, 21 Jul 2020 08:17:49 -0700

chiark-utils (6.1.2) unstable; urgency=medium

  * git-cache-proxy: Recheck gc timestamp after taking lock
  * git-cache-proxy: Tolerate failure to write gc stamp
  * git-cache-proxy: Do not update timestamp due to housekeeping gc

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Wed, 12 Feb 2020 16:10:02 +0000

chiark-utils (6.1.2~citrix2~) unstable; urgency=medium

  * git-cache-proxy: Move $SIG{ALRM} setting to global
  * git-cache-proxy: Abandon fetch attempt if client disconnects
  * git-cache-proxy: Use open-coded fork/exec for git gc

 -- Ian Jackson <ian.jackson@citrix.com>  Tue, 11 Feb 2020 18:34:58 +0000

chiark-utils (6.1.1) unstable; urgency=medium

  * fishdescriptor: Use Python "errno" module
  * git-cache-proxy: Periodically run `git gc --quiet'
  * git-cache-proxy: Document some options in the head comment

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Tue, 11 Feb 2020 16:37:31 +0000

chiark-utils (6.0.5~citrix1) unstable; urgency=medium

  * fishdescriptor: Use Python "errno" module, not "os.errno"
    (which is not in modern python3 apparently).

 -- Ian Jackson <ian.jackson@citrix.com>  Fri, 20 Dec 2019 17:50:02 +0000

chiark-utils (6.1.0~iwj1) unstable; urgency=medium

  * really: Writeability of rc.local suffices, too.
    (Non-sysvinit compatibility.)

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 12 Aug 2019 10:59:02 +0100

chiark-utils (6.1.0) unstable; urgency=medium

  * Non-maintainer upload.
    - Maintainer has approved this via personal communication.
  * git-branchmove: rewrite in perl (Closes: #914398, #914399)
    - Add dependencies on libgit-wrapper-perl, libtry-tiny-perl to
      bin:chiark-scripts.
  * git-branchmove: new --detach feature.

 -- Sean Whitton <spwhitton@spwhitton.name>  Sat, 07 Dec 2019 14:10:26 -0700

chiark-utils (6.0.4) unstable; urgency=medium

  * sync-accounts: Fix perl syntax error.  Closes:#865985.
  * changelog: Document bug number for bugfix in 6.0.4~citrix1.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Fri, 12 Apr 2019 11:48:00 +0100

chiark-utils (6.0.4~citrix1) unstable; urgency=medium

  * fishdescriptor: cast __errno_location correctly.  Closes:#926858.

 -- Ian Jackson <ian.jackson@citrix.com>  Mon, 08 Apr 2019 17:03:47 +0100

chiark-utils (6.0.3) unstable; urgency=medium

  * Upload to Debian unstable.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Tue, 20 Nov 2018 18:25:53 +0000

chiark-utils (6.0.3~citrix1) unstable; urgency=medium

  * git-branchmove: Avoid transporting tags about
  * fishdescriptor: Cope if donor is in a chroot without /proc
  * fishdescriptor: Cope if only donor has /run/user/UID

 -- Ian Jackson <ian.jackson@citrix.com>  Mon, 19 Nov 2018 16:11:55 +0000

chiark-utils (6.0.2) unstable; urgency=medium

  * Rebuild with arch:all packages included in the upload.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 11 Jun 2018 14:35:08 +0100

chiark-utils (6.0.1) unstable; urgency=low

  fishdescriptor:
  * Fix numerous bugs, affecting the `exec' feature with nontrivial fd
    specifications.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 11 Jun 2018 14:31:24 +0100

chiark-utils (6.0.0) unstable; urgency=medium

  xbatmon-simple:
  * Minor improvements to the .txt doc.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Fri, 20 Apr 2018 16:56:04 +0100


cprogs/summer.c
/*
 * summer - program for summarising (with md5 checksums) filesystem trees
 *
 * usage:
 *    cat startpoints.list | summer >data.list
 *    summer startpoints... >data.list
 *  prints md5sum of data-list to stderr
 */
/*
 * Copyright (C) 2003,2006-2007 Ian Jackson <ian@davenant.greenend.org.uk>
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3,
 * or (at your option) any later version.
 *
 * This is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this file; if not, consult the Free Software
 * Foundation's website at www.fsf.org, or the GNU Project website at
 * www.gnu.org.
 */

#define _GNU_SOURCE

#include <search.h>
#include <dirent.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdio.h>
#include <inttypes.h>
#include <string.h>
#include <errno.h>
#include <stdarg.h>
#include <limits.h>
#include <assert.h>
#include <stdlib.h>

#include "nettle/md5-compat.h"

#define MAXFN 2048
#define MAXDEPTH 1024
#define CSUMXL 32

static int quiet=0, hidectime=0, hideatime=0, hidemtime=0;
static int hidedirsize=0, hidelinkmtime=0, hidextime=0, onefilesystem=0;
static int filenamefieldsep=' ';
static FILE *errfile;

#define nodeflag_fsvalid       1u

static void malloc_fail(void) { perror("summer: alloc failed"); exit(12); }

static void *mmalloc(size_t sz) {
  void *r;
  r= malloc(sz);  if (!r) malloc_fail();
  return r;
}

static void *mrealloc(void *p, size_t sz) {
  void *r;
  r= realloc(p, sz);  if (!r && sz) malloc_fail();
  return r;
}

static void fn_escaped(FILE *f, const char *fn) {
  int c;
  while ((c= *fn++)) {
    if (c>=33 && c<=126 && c!='\\') putc(c,f);
    else fprintf(f,"\\x%02x",(int)(unsigned char)c);
  }
}

static void add_pr(int *pr, int printf_ret) {
  if (printf_ret == EOF) return;
  *pr += printf_ret;
}

static void vproblemx(const char *path, int padto, int per,
		      const char *fmt, va_list al) {
  int e=errno, pr=0;
  
  if (errfile==stderr) fputs("summer: error: ",stderr);
  else add_pr(&pr, fprintf(errfile,"\\["));
  
  add_pr(&pr, vfprintf(errfile,fmt,al));
  if (per) add_pr(&pr, fprintf(errfile,": %s",strerror(e)));

  if (errfile==stderr) {
    fputs(": ",stderr);
    fn_escaped(stderr,path);
    fputc('\n',stderr);
    exit(2);
  }

  add_pr(&pr, printf("]"));

  while (pr++ < padto)
    putchar(' ');
}  

static void problem_e(const char *path, int padto, const char *fmt, ...) {
  va_list(al);
  va_start(al,fmt);
  vproblemx(path,padto,1,fmt,al);
  va_end(al);
}

static void problem(const char *path, int padto, const char *fmt, ...) {
  va_list(al);
  va_start(al,fmt);
  vproblemx(path,padto,0,fmt,al);
  va_end(al);
}

static void csum_file(const char *path) {
  FILE *f;
  MD5_CTX mc;
  char db[65536];
  unsigned char digest[16];
  size_t r;
  int i;

  f= fopen(path,"rb");
  if (!f) { problem_e(path,sizeof(digest)*2,"open"); return; }
  
  MD5Init(&mc);
  for (;;) {
    r= fread(db,1,sizeof(db),f);
    if (ferror(f)) {
      problem_e(path,sizeof(digest)*2,"read");
      fclose(f); return;
    }
    if (!r) { assert(feof(f)); break; }
    MD5Update(&mc,db,r);
  }
  MD5Final(digest,&mc);
  if (fclose(f)) { problem_e(path,sizeof(digest)*2,"close"); return; }

  for (i=0; i<sizeof(digest); i++)
    printf("%02x", digest[i]);
}

static void csum_dev(int cb, const struct stat *stab) {
  printf("%c 0x%08lx %3lu %3lu %3lu %3lu    ", cb,
	 (unsigned long)stab->st_rdev,
	 ((unsigned long)stab->st_rdev & 0x0ff000000U) >> 24,
	 ((unsigned long)stab->st_rdev & 0x000ff0000U) >> 16,
	 ((unsigned long)stab->st_rdev & 0x00000ff00U) >> 8,
	 ((unsigned long)stab->st_rdev & 0x0000000ffU) >> 0);
}

static void csum_str(const char *s) {
  printf("%-*s", CSUMXL, s);
}

static void linktargpath(const char *linktarg) {
  printf(" -> ");
  fn_escaped(stdout, linktarg);
}

static void pu10(void) { printf(" %10s", "?"); }

#define PTIME(stab, memb)  ((stab) ? ptime((stab), (stab)->memb) : pu10())

static void ptime(const struct stat *stab, uint64_t val) {
  const char *instead;

  if (!hidextime) goto justprint;
  else if (S_ISCHR(stab->st_mode)) instead= "char";
  else if (S_ISBLK(stab->st_mode)) instead= "block";
  else if (S_ISLNK(stab->st_mode)) instead= "link";
  else if (S_ISSOCK(stab->st_mode)) instead= "sock";
  else if (S_ISFIFO(stab->st_mode)) instead= "pipe";
  else {
  justprint:
    printf(" %10" PRIu64 "", val);
    return;
  }

  printf(" %10s",instead);
}

struct hardlink {
  dev_t dev;
  ino_t ino;
  char path[1];
};
static void *hardlinks;

static int hardlink_compar(const void *av, const void *bv) {
  const struct hardlink *a=av, *b=bv;
  if (a->ino != b->ino) return b->ino - a->ino;
  return b->dev - a->dev;
}

static void recurse(const char *path, unsigned nodeflags, dev_t fs);

static void node(const char *path, unsigned nodeflags, dev_t fs) {
  char linktarg[MAXFN+1];
  struct hardlink *foundhl;
  const struct stat *stab;
  struct stat stabuf;
  int r, mountpoint=0;

  r= lstat(path, &stabuf);
  stab= r ? 0 : &stabuf;

  foundhl= 0;
  if (stab && stab->st_nlink>1) {
    struct hardlink *newhl, **foundhl_node;
    newhl= mmalloc(sizeof(*newhl) + strlen(path));
    newhl->dev= stab->st_dev;
    newhl->ino= stab->st_ino;
    foundhl_node= tsearch(newhl, &hardlinks, hardlink_compar);
    if (!foundhl_node) malloc_fail();
    foundhl= *foundhl_node;
    if (foundhl!=newhl) {
      free(newhl); /* hardlink to an earlier object */
    } else {
      foundhl= 0; /* new object with link count>1 */
      strcpy(newhl->path, path);
    }
  }

  if (stab) {
    if ((nodeflags & nodeflag_fsvalid) && stab->st_dev != fs)
      mountpoint= 1;
    fs= stab->st_dev;
    nodeflags |= nodeflag_fsvalid;
  }

  if (!stab) problem_e(path,CSUMXL,"inaccessible");
  else if (foundhl) csum_str("hardlink");
  else if (S_ISREG(stab->st_mode)) csum_file(path);
  else if (S_ISCHR(stab->st_mode)) csum_dev('c',stab);
  else if (S_ISBLK(stab->st_mode)) csum_dev('b',stab);
  else if (S_ISFIFO(stab->st_mode)) csum_str("pipe");
  else if (S_ISLNK(stab->st_mode)) csum_str("symlink");
  else if (S_ISSOCK(stab->st_mode)) csum_str("sock");
  else if (S_ISDIR(stab->st_mode)) csum_str(mountpoint ? "mountpoint" : "dir");
  else problem(path,CSUMXL,"badobj: 0x%lx", (unsigned long)stab->st_mode);

  if (stab && S_ISLNK(stab->st_mode)) {
    r= readlink(path, linktarg, sizeof(linktarg)-1);
    if (r==sizeof(linktarg)) { problem(path,-1,"readlink too big"); r=-1; }
    else if (r<0) { problem_e(path,-1,"readlink"); }
    else assert(r<sizeof(linktarg));

    if (r<0) strcpy(linktarg,"\\?");
    else linktarg[r]= 0;
  }

  if (stab) {
    if (S_ISDIR(stab->st_mode) && hidedirsize)
      printf(" %10s","dir");
    else
      printf(" %10llu", 
	     (unsigned long long)stab->st_size);

    printf(" %4o %10ld %10ld",
	   (unsigned)stab->st_mode & 07777U,
	   (unsigned long)stab->st_uid,
	   (unsigned long)stab->st_gid);
  } else {
    printf(" %10s %4s %10s %10s", "?","?","?","?");
  }

  if (!hideatime)
    PTIME(stab, st_atime);

  if (!hidemtime) {
    if (stab && S_ISLNK(stab->st_mode) && hidelinkmtime)
      printf(" %10s","link");
    else
      PTIME(stab, st_mtime);
  }

  if (!hidectime)
    PTIME(stab, st_ctime);

  putchar(filenamefieldsep);
  fn_escaped(stdout, path);

  if (foundhl) linktargpath(foundhl->path);
  if (stab && S_ISLNK(stab->st_mode)) linktargpath(linktarg);

  putchar('\n');

  if (ferror(stdout)) { perror("summer: stdout"); exit(12); }

  if (stab && S_ISDIR(stab->st_mode) && !(mountpoint && onefilesystem))
    recurse(path, nodeflags, fs);
}

static void process(const char *startpoint) {
  if (!quiet)
    fprintf(stderr,"summer: processing: %s\n",startpoint);
  node(startpoint, 0,0);
  tdestroy(hardlinks,free);
  hardlinks= 0;
}

static int recurse_maxlen;

static int recurse_filter(const struct dirent *de) {
  int l;
  if (de->d_name[0]=='.' &&
      (de->d_name[1]==0 ||
       (de->d_name[1]=='.' &&
	de->d_name[2]==0)))
    return 0;
  l= strlen(de->d_name);
  if (l > recurse_maxlen) recurse_maxlen= l;
  return 1;
}

static int recurse_compar(const struct dirent **a, const struct dirent **b) {
  return strcmp((*a)->d_name, (*b)->d_name);
}

static void recurse(const char *path_or_buf, unsigned nodeflags, dev_t fs) {
  static char *buf;
  static int buf_allocd;
  
  struct dirent **namelist, *const *de;
  const char *path_or_0= path_or_buf==buf ? 0 : path_or_buf;
  int nentries, pathl, esave, buf_want, i;

  pathl= strlen(path_or_buf);
  recurse_maxlen= 2;
  nentries= scandir(path_or_buf, &namelist, recurse_filter, recurse_compar);
  esave= errno;
  
  buf_want= pathl+1+recurse_maxlen+1;
  if (buf_want > buf_allocd) {
    buf= mrealloc(buf, buf_want);
    buf_allocd= buf_want;
  }
  /* NOTE that path_or_buf is invalid after this point because
   * it might have been realloc'd ! */
  if (path_or_0) strcpy(buf,path_or_0);

  buf[pathl]= '/';
  pathl++;
  if (nentries < 0) {
    buf[pathl]= 0;  errno= esave;
    problem_e(buf,CSUMXL+72,"scandir failed");
    fn_escaped(stdout,buf);  putchar('\n');
    return;
  }
  for (i=0, de=namelist; i<nentries; i++, de++) {
    strcpy(buf+pathl, (*de)->d_name);
    node(buf, nodeflags, fs);
    free(*de);
  }
  free(namelist);
}

static void from_stdin(void) {
  char buf[MAXFN+2];
  char *s;
  int l;

  if (!quiet)
    fprintf(stderr, "summer: processing stdin lines as startpoints\n");
  for (;;) {
    s= fgets(buf,sizeof(buf),stdin);
    if (ferror(stdin)) { perror("summer: stdin"); exit(12); }
    if (!s) { if (feof(stdin)) return; else abort(); }
    l= strlen(buf);
    assert(l>0);
    if (buf[l-1]!='\n') { fprintf(stderr,"summer: line too long\n"); exit(8); }
    buf[l-1]= 0;
    process(buf);
  }
}

int main(int argc, const char *const *argv) {
  const char *arg;
  int c;

  errfile= stderr;
  
  while ((arg=argv[1]) && *arg++=='-') {
    while ((c=*arg++)) {
      switch (c) {
      case 'h':
	fprintf(stderr,
		"summer: usage: summer startpoint... >data.list\n"
		"               cat startpoints.list | summer >data.list\n");
	exit(8);
      case 'q':
	quiet= 1;
	break;
      case 't':
	filenamefieldsep= '\t';
	break;
      case 'D':
	hidedirsize= 1;
	break;
      case 'b':
	hidelinkmtime= 1;
	break;
      case 'B':
	hidextime= 1;
	break;
      case 'x':
	onefilesystem= 1;
	break;
      case 'C':
	hidectime= 1;
	break;
      case 'A':
	hideatime= 1;
	break;
      case 'M':
	hidemtime= 1;
	break;
      case 'f':
	errfile= stdout;
	break;
      default:
	fprintf(stderr,"summer: bad usage, try -h\n");
	exit(8);
      }
    }
    argv++;
  }

  if (!argv[1]) {
    from_stdin();
  } else {
    if (!quiet)
      fprintf(stderr, "summer: processing command line args as startpoints\n");
    while ((arg=*++argv)) {
      process(arg);
    }
  }
  if (ferror(stdout) || fclose(stdout)) {
    perror("summer: stdout (at end)"); exit(12);
  }
  if (!quiet)
    fputs("summer: done.\n", stderr);
  return 0;
}
